README.md
# @ianwalter/requester
> A convenience wrapper around the Node.js [http.request][nodeUrl] API

[![npm page][npmImage]][npmUrl]
[![CI][ciImage]][ciUrl]

## Installation

```console
yarn add @ianwalter/requester
```

## Usage

Simple example GET request:

```js
const { requester } = require('@ianwalter/requester')

const response = await requester.get('http://example.com/api/v1/cats')
```

Example POST request with custom base options:

```js
const { Requester } = require('@ianwalter/requester')

const requester = new Requester({ baseUrl: 'http://example.com/api' })

const response = await requester.post('/v1/cats', { name: 'Nibblet' })
```

Example GET request with a custom request header:

```js
const { requester } = require('@ianwalter/requester')

const options = { headers: { authorization: 'Bearer abc123' } }
const response = await requester.get('http://example.com/api/v1/cats', options)
```

## License

Hippocratic License - See [LICENSE][licenseUrl]

&nbsp;

Created by [Ian Walter](https://ianwalter.dev)

[nodeUrl]: https://nodejs.org/api/http.html#http_http_request_url_options_callback
[npmImage]: https://img.shields.io/npm/v/@ianwalter/requester.svg
[npmUrl]: https://www.npmjs.com/package/@ianwalter/requester
[ciImage]: https://github.com/ianwalter/requester/workflows/CI/badge.svg
[ciUrl]: https://github.com/ianwalter/requester/actions
[licenseUrl]: https://github.com/ianwalter/requester/blob/main/LICENSE


CHANGELOG.md
# @ianwalter/requester

## 2.5.0

### Minor Changes

- dc15cc5: Automatically following redirects

## 2.4.3

### Patch Changes

- 5367870: Fixing empty body issue in HttpError

## 2.4.1

### Patch Changes

- b90e5c3: Quick fix for body JSON.stringify Date issue

## 2.4.0

### Minor Changes

- bccf908: Adding request url, options to response


tests.js
const { test } = require('@ianwalter/bff')
const { createKoaServer } = require('@ianwalter/test-server')
const { requester, Requester } = require('.')

test('GET request for empty response', async ({ expect }) => {
  const server = await createKoaServer()
  server.use(ctx => (ctx.status = 204))
  const response = await requester.get(server.url)
  expect(response.ok).toBe(true)
  expect(response.statusCode).toBe(204)
  expect(response.body).toBe(undefined)
  await server.close()
})

test('GET request for text', async ({ expect }) => {
  const server = await createKoaServer()
  server.use(ctx => (ctx.body = 'test'))
  const response = await requester.get(server.url)
  expect(response.ok).toBe(true)
  expect(response.statusCode).toBe(200)
  expect(response.body).toBe('test')
  await server.close()
})

test('GET request for JSON', async ({ expect }) => {
  const server = await createKoaServer()
  const body = { message: 'test' }
  server.use(ctx => (ctx.body = body))
  const response = await requester.get(server.url)
  expect(response.ok).toBe(true)
  expect(response.statusCode).toBe(200)
  expect(response.body).toEqual(body)
  await server.close()
})

test('POST request with JSON', async ({ expect }) => {
  const server = await createKoaServer()
  const body = { chef: 'Sanchez' }
  server.use(ctx => {
    ctx.status = ctx.request.body.chef === body.chef ? 201 : 400
    ctx.body = ctx.request.body
  })
  const response = await requester.post(server.url, { body })
  expect(response.ok).toBe(true)
  expect(response.statusCode).toBe(201)
  expect(response.body).toEqual(body)
  await server.close()
})

test('Unauthorized GET request', async ({ expect }) => {
  const server = await createKoaServer()
  server.use(ctx => (ctx.status = 401))
  try {
    await requester.get(server.url)
  } catch (err) {
    expect(err.response.ok).toBe(false)
    expect(err.response.statusCode).toBe(401)
    expect(err.response.body).toBe('Unauthorized')
  }
  await server.close()
})

test('Bad Request with shouldThrow = false', async ({ expect }) => {
  const requester = new Requester({ shouldThrow: false })
  const server = await createKoaServer()
  const body = { message: 'Ungodly gorgeous, buried in a chorus' }
  server.use(ctx => {
    ctx.status = 400
    ctx.body = body
  })
  const response = await requester.get(server.url)
  expect(response.ok).toBe(false)
  expect(response.statusCode).toBe(400)
  expect(response.body).toEqual(body)
  await server.close()
})

test('HTTPS request', async ({ expect }) => {
  const requestbin = 'https://en5femwzf9sry.x.pipedream.net'
  const body = { success: true }
  let response = await requester.get(requestbin)
  expect(response.statusCode).toBe(200)
  expect(response.body).toEqual(body)
  response = await requester.post(requestbin, { body })
  expect(response.statusCode).toBe(200)
  expect(response.body).toEqual(body)
})


tests/tests.js
const querystring = require('querystring')
const { test } = require('@ianwalter/bff')
const { createApp } = require('@ianwalter/nrg')
const { requester, Requester } = require('..')

test('GET request for empty response', async ({ expect }) => {
  const app = createApp({ log: false })
  app.use(ctx => (ctx.status = 204))
  const server = await app.serve()
  try {
    const response = await requester.get(server.url)
    expect(response.ok).toBe(true)
    expect(response.statusCode).toBe(204)
    expect(response.body).toBe(undefined)
  } finally {
    await server.close()
  }
})

test('GET request for text', async ({ expect }) => {
  const app = createApp({ log: false })
  app.use(ctx => (ctx.body = 'test'))
  const server = await app.serve()
  try {
    const response = await requester.get(server.url)
    expect(response.ok).toBe(true)
    expect(response.statusCode).toBe(200)
    expect(response.body).toBe('test')
  } finally {
    await server.close()
  }
})

test('GET request for JSON', async ({ expect }) => {
  const app = createApp({ log: false })
  const body = { message: 'test' }
  app.use(ctx => (ctx.body = body))
  const server = await app.serve()
  try {
    const response = await requester.get(server.url)
    expect(response.ok).toBe(true)
    expect(response.statusCode).toBe(200)
    expect(response.body).toEqual(body)
  } finally {
    await server.close()
  }
})

test('POST request with JSON', async ({ expect }) => {
  const app = createApp({ log: false })
  const body = { chef: 'Sanchez', born: new Date() }
  app.use(ctx => {
    ctx.status = ctx.request.body.chef === body.chef ? 201 : 400
    ctx.body = ctx.request.body
  })
  const server = await app.serve()
  try {
    const response = await requester.post(server.url, { body })
    expect(response.ok).toBe(true)
    expect(response.statusCode).toBe(201)
    expect(response.body).toEqual({ ...body, born: body.born.toISOString() })
  } finally {
    await server.close()
  }
})

test('POST request for form data', async ({ expect }) => {
  const body = { artist: 'Peter Bjorn and John', song: 'Music' }
  const app = createApp({ log: false })
  app.use(ctx => {
    ctx.set('content-type', 'application/x-www-form-urlencoded')
    ctx.body = querystring.stringify(body)
  })
  const server = await app.serve()
  try {
    const response = await requester.post(server.url)
    expect(response.ok).toBe(true)
    expect(response.statusCode).toBe(200)
    expect(response.body).toEqual(body)
  } finally {
    await server.close()
  }
})

test('Unauthorized GET request', async ({ expect }) => {
  const app = createApp({ log: false })
  app.use(ctx => (ctx.status = 401))
  const server = await app.serve()
  try {
    await requester.get(server.url)
  } catch (err) {
    expect(err.response.ok).toBe(false)
    expect(err.response.statusCode).toBe(401)
    expect(err.response.body).toBe('Unauthorized')
  } finally {
    await server.close()
  }
})

test('Bad Request with shouldThrow = false', async ({ expect }) => {
  const app = createApp({ log: false })
  const requester = new Requester({ shouldThrow: false })
  const body = { message: 'Ungodly gorgeous, buried in a chorus' }
  app.use(ctx => {
    ctx.status = 400
    ctx.body = body
  })
  const server = await app.serve()
  try {
    const response = await requester.get(server.url)
    expect(response.ok).toBe(false)
    expect(response.statusCode).toBe(400)
    expect(response.body).toEqual(body)
  } finally {
    await server.close()
  }
})

test('HTTPS request', async ({ expect }) => {
  const requestbin = 'https://en5femwzf9sry.x.pipedream.net'
  const body = { success: true }
  let response = await requester.get(requestbin)
  expect(response.statusCode).toBe(200)
  expect(response.body).toEqual(body)
  response = await requester.post(requestbin, { body })
  expect(response.statusCode).toBe(200)
  expect(response.body).toEqual(body)
})

test('Get request options from response', async t => {
  const app = createApp({ log: false })
  app.use(ctx => (ctx.status = 204))
  const server = await app.serve()
  const headers = { 'x-test': 123 }
  try {
    const { statusCode, request } = await requester.get(server.url, { headers })
    t.expect(statusCode).toBe(204)
    t.expect(request.options.headers['x-test']).toEqual(headers['x-test'])
  } finally {
    await server.close()
  }
})

// TODO:
// test('GET gzipped response', async ({ expect }) => {
//   const largeJsonBody = require('./largeJsonBody.json')
//   const server = await createExpressServer()
//   const headers = { 'accept-encoding': 'gzip, deflate, br' }
//   server.use((req, res) => res.json(largeJsonBody))
//   const response = await requester.get(server.url, { headers })
//   expect(typeof response.body).toBe('object')
//   await server.close()
// })


tests/tests.mjs
import querystring from 'querystring'
import { test } from '@ianwalter/bff'
import nrg from '@ianwalter/nrg'
import rqr from '../index.js'

const { requester, Requester } = rqr

test('GET request for empty response', async t => {
  const app = nrg.createApp({ log: false })
  app.use(ctx => (ctx.status = 204))
  const server = await app.serve()
  try {
    const response = await requester.get(server.url)
    t.expect(response.ok).toBe(true)
    t.expect(response.statusCode).toBe(204)
    t.expect(response.body).toBe(undefined)
  } finally {
    await server.close()
  }
})

test('GET request for text', async t => {
  const app = nrg.createApp({ log: false })
  app.use(ctx => (ctx.body = 'test'))
  const server = await app.serve()
  try {
    const response = await requester.get(server.url)
    t.expect(response.ok).toBe(true)
    t.expect(response.statusCode).toBe(200)
    t.expect(response.body).toBe('test')
  } finally {
    await server.close()
  }
})

test('GET request for JSON', async t => {
  const app = nrg.createApp({ log: false })
  const body = { message: 'test' }
  app.use(ctx => (ctx.body = body))
  const server = await app.serve()
  try {
    const response = await requester.get(server.url)
    t.expect(response.ok).toBe(true)
    t.expect(response.statusCode).toBe(200)
    t.expect(response.body).toEqual(body)
  } finally {
    await server.close()
  }
})

test('POST request with JSON', async t => {
  const app = nrg.createApp({ log: false })
  const body = { chef: 'Sanchez', born: new Date() }
  app.use(ctx => {
    ctx.status = ctx.request.body.chef === body.chef ? 201 : 400
    ctx.body = ctx.request.body
  })
  const server = await app.serve()
  try {
    const response = await requester.post(server.url, { body })
    t.expect(response.ok).toBe(true)
    t.expect(response.statusCode).toBe(201)
    t.expect(response.body).toEqual({ ...body, born: body.born.toISOString() })
  } finally {
    await server.close()
  }
})

test('POST request for form data', async t => {
  const body = { artist: 'Peter Bjorn and John', song: 'Music' }
  const app = nrg.createApp({ log: false })
  app.use(ctx => {
    ctx.set('content-type', 'application/x-www-form-urlencoded')
    ctx.body = querystring.stringify(body)
  })
  const server = await app.serve()
  try {
    const response = await requester.post(server.url)
    t.expect(response.ok).toBe(true)
    t.expect(response.statusCode).toBe(200)
    t.expect(response.body).toEqual(body)
  } finally {
    await server.close()
  }
})

test('Unauthorized GET request', async t => {
  const app = nrg.createApp({ log: false })
  app.use(ctx => (ctx.status = 401))
  const server = await app.serve()
  try {
    await requester.get(server.url)
  } catch (err) {
    t.expect(err.response.ok).toBe(false)
    t.expect(err.response.statusCode).toBe(401)
    t.expect(err.response.body).toBe('Unauthorized')
  } finally {
    await server.close()
  }
})

test('Bad Request with shouldThrow = false', async t => {
  const app = nrg.createApp({ log: false })
  const requester = new Requester({ shouldThrow: false })
  const body = { message: 'Ungodly gorgeous, buried in a chorus' }
  app.use(ctx => {
    ctx.status = 400
    ctx.body = body
  })
  const server = await app.serve()
  try {
    const response = await requester.get(server.url)
    t.expect(response.ok).toBe(false)
    t.expect(response.statusCode).toBe(400)
    t.expect(response.body).toEqual(body)
  } finally {
    await server.close()
  }
})

test('HTTPS request', async t => {
  const requestbin = 'https://en5femwzf9sry.x.pipedream.net'
  const body = { success: true }
  let response = await requester.get(requestbin)
  t.expect(response.statusCode).toBe(200)
  t.expect(response.body).toEqual(body)
  response = await requester.post(requestbin, { body })
  t.expect(response.statusCode).toBe(200)
  t.expect(response.body).toEqual(body)
})

test('Redirect', async t => {
  const app = nrg.createApp({ log: false })

  app.get('/', ctx => {
    ctx.set('location', server.url + '/a')
    ctx.status = 301
  })

  app.get('/a', ctx => (ctx.body = 'Success!'))

  const server = await app.serve()

  try {
    const response = await requester.get(server.url)
    t.expect(response.statusCode).toBe(200)
    t.expect(response.body).toBe('Success!')
  } finally {
    await server.close()
  }
})

test('Get request options from response', async t => {
  const app = nrg.createApp({ log: false })
  app.use(ctx => (ctx.status = 204))
  const server = await app.serve()
  const headers = { 'x-test': 123 }
  try {
    const { statusCode, request } = await requester.get(server.url, { headers })
    t.expect(statusCode).toBe(204)
    t.expect(request.options.headers['x-test']).toEqual(headers['x-test'])
  } finally {
    await server.close()
  }
})

// TODO:
// test('GET gzipped response', async ({ expect }) => {
//   const largeJsonBody = require('./largeJsonBody.json')
//   const server = await createExpressServer()
//   const headers = { 'accept-encoding': 'gzip, deflate, br' }
//   server.use((req, res) => res.json(largeJsonBody))
//   const response = await requester.get(server.url, { headers })
//   expect(typeof response.body).toBe('object')
//   await server.close()
// })


index.js
const http = require('http')
const https = require('https')
const { URL } = require('url')
const zlib = require('zlib')
const querystring = require('querystring')
const { createLogger } = require('@generates/logger')
const { version } = require('./package.json')
const { merge } = require('@generates/merger')

const methods = [
  'get',
  'post',
  'put',
  'delete'
]
const headers = {
  'user-agent': `@ianwalter/requester/${version}`
}
const urlencoded = 'application/x-www-form-urlencoded'
const defaults = {
  shouldThrow: true,
  logLevel: 'info',
  headers,
  timeout: 60000
}
const logger = createLogger({ level: 'info', namespace: 'requester' })

class HttpError extends Error {
  constructor (response) {
    super(response?.statusText || 'Request failed')
    this.name = this.constructor.name
    this.response = response

    // Remove rawBody so that it doesn't get printed and flood the console.
    delete this.response?.body?.rawBody
  }
}

class Requester {
  constructor (options) {
    // Set the base options for the requester instance.
    this.options = merge({}, defaults, options)

    // Add convenience methods to the requester instance.
    methods.forEach(method => {
      this[method] = async (url, opts) => this.request(url, { method, ...opts })
    })
  }

  shapeRequest (options) {
    // If a body object or array was passed, automatically stringify it and add
    // a JSON Content-Type header.
    if (options.body && typeof options.body === 'object') {
      logger.debug('Request body is JSON', { body: options.body })
      options.headers['content-type'] = 'application/json'
      options.body = JSON.stringify(options.body)
      options.headers['content-length'] = `${Buffer.byteLength(options.body)}`
    }
  }

  shapeResponse (response) {
    // Add the .ok convenience property.
    response.ok = response.statusCode < 400 && response.statusCode > 199

    if (response.body) {
      // Decompress the response if it has a content-encoding header indicating
      // the body is encoded using gzip, deflate, or brotli.
      const encoding = response.headers && response.headers['content-encoding']
      if (encoding) {
        if (encoding === 'gzip') {
          logger.debug('Response body is encoded using gzip')
          response.body = zlib.gunzipSync(response.body)
        } else if (encoding === 'br') {
          logger.debug('Response body is encoded using brotli')
          response.body = zlib.brotliDecompressSync(response.body)
        } else if (encoding === 'deflate') {
          logger.debug('Response body is encoded using deflate')
          response.body = zlib.deflateSync(response.body)
        }
      }

      // If the response content type is text or JSON, set the body Buffer to
      // rawBody and the stringified version of it to body on the response.
      const contentType = response.headers && response.headers['content-type']
      const isJson = contentType && contentType.includes('application/json')
      const isUrlEncoded = contentType && contentType.includes(urlencoded)
      if (
        !contentType ||
        (contentType && contentType.includes('text/')) ||
        isJson ||
        isUrlEncoded
      ) {
        response.rawBody = response.body
        response.body = response.body.toString()
      }

      // Automatically parse the response body as JSON if the Content-Type
      // header is application/json.
      if (isJson) {
        logger.debug('Response body is JSON')
        try {
          response.body = JSON.parse(response.body)
        } catch (err) {
          logger.error(err)
        }
      } else if (isUrlEncoded) {
        response.body = querystring.parse(response.body)
      }
    }
  }

  request (url, { body, ...options }) {
    // Combine the base options and argument options into a single object.
    options = merge({ body }, this.options, options)

    // If a base URL has been configured, use it to build the complete URL.
    if (options.baseUrl) url = new URL(url, options.baseUrl).toString()

    // Automatically add request headers based on the request body.
    this.shapeRequest(options)
    logger.debug('Request', { url, options })

    return new Promise((resolve, reject) => {
      // Create the request.
      const client = url.indexOf('http://') === 0 ? http : https
      const request = client.request(url, options)

      // If an error event was received, reject the returned Promise.
      request.once('error', err => {
        err.request = { url, options }
        reject(err)
      })

      request.once('response', response => {
        const bodyChunks = []

        const { statusCode, statusText, headers } = response
        logger.debug('Response', { statusCode, statusText, headers })

        if ([301, 302, 307, 308].includes(response.statusCode)) {
          const { location } = response.headers
          return resolve(this.request(location, { body, ...options }))
        }

        // Listen to the data event to receive the response body as one or more
        // buffers and collect them into the bodyChunks array.
        response.on('data', data => {
          if (this.options.logLevel === 'debug') {
            logger.debug('Response data event', data.toString())
          }
          bodyChunks.push(data)
        })

        // When the response is complete, resolve the returned Promise with the
        // response.
        response.on('end', () => {
          logger.debug('Response end event')

          // Set the response body to a single Buffer by concatenating the
          // chunks in the bodyChunks array.
          if (bodyChunks.length) response.body = Buffer.concat(bodyChunks)

          // Shape the response based on the received response headers.
          this.shapeResponse(response)

          // Add a request object to the repsonse that has the requested URL
          // and options.
          response.request = { url, options }

          if (options.shouldThrow && !response.ok) {
            reject(new HttpError(response))
          } else {
            resolve(response)
          }
        })
      })

      request.once('timeout', () => {
        logger.debug('Request timeout event')
        request.abort()
      })

      // If a request body was passed, write it to the request stream.
      if (options.body) request.write(options.body)

      // Execute the request by ending the stream.
      request.end()
    })
  }
}

module.exports = { Requester, requester: new Requester(), HttpError }
